### PR TITLE
Define a explicit and generic socket<->transport contract

### PR DESCRIPTION
**Pull request by josevalim**
This PR is still work in progress but I am submitting it early for discussion.

The main goal of this PR is to define a contract between transports and sockets. This brings two very interesting features:

1. Developers can use Phoenix.Socket as a generic communication layer, no longer coupled to channels. For example, you can implement your own callbacks such as a simple ping/pong communication layer and Phoenix will make sure it works for both websockets AND long polling.

2. Implementing a transport is much much much easier as it only needs to communicate with the socket and is no longer required to worry about channels.

Basically, the whole channels management is only one possible implementation of the socket<->transport contract. You can replace this implementation altogether by not using Phoenix.Socket and defining your own callbacks instead.

/cc @sasa1977 

### TODO

 * [ ] Non-channel based integration tests
 * [x] Deprecate transport macro in sockets
 * [x] Phoenix.Socket and adapter docs
 * [x] Update guides (transport)
 * [ ] CHANGELOG (transport/3, Phoenix.Socket.Transport, serializers, :handler)

### CODE REVIEW

**Comment by josevalim** [comment]
Btw, the PR has an increase in the lines of code but that's because of deprecation warnings and extra documentation. Overall it should less LOC since we removed duplication in the serializers and existing transports.

File: lib/phoenix/channel/server.ex [hunk 1]
```diff
@@ -183,13 +218,14 @@ defmodule Phoenix.Channel.Server do
 
   @doc false
   def init({socket, auth_payload, parent, ref}) do
+    _ = Process.monitor(socket.transport_pid)
```
Comment (Gazler) [inline lib/phoenix/channel/server.ex]: Is the match here deliberate?

File: lib/phoenix/transports/long_poll.ex [hunk 2]
```diff
@@ -261,18 +230,21 @@ defmodule Phoenix.Transports.LongPoll do
...
   end
 
-  defp code_reload(conn, opts, endpoint) do
-    reload? = Keyword.get(opts, :code_reloader, endpoint.config(:code_reloader))
-    if reload?, do: Phoenix.CodeReloader.reload!(endpoint)
+  defp status_token_messages_json(conn, token, messages) do
+    status = conn.status || 200
+    messages = [?[, Enum.intersperse(messages, ?,), ?]]
+    token = Phoenix.json_library().encode_to_iodata!(token)
+    send_json(conn, "{\"status\":#{status},\"token\":#{token},\"messages\":#{messages}}")
```
Comment (Gazler) [inline lib/phoenix/transports/long_poll.ex]: Is there any reason not to use a map here and `Phoenix.json_library().encode` the whole thing?

File: mix.exs [hunk 3]
```diff
@@ -178,9 +178,7 @@ defmodule Phoenix.MixProject do
   end
 
   defp aliases do
-    [
-      "docs": ["docs", &generate_js_docs/1]
-    ]
+    ["docs": ["docs", &generate_js_docs/1]]
```
Comment (Gazler) [inline mix.exs]: Was this change intentional?

File: lib/phoenix/socket/transport.ex [hunk 4]
```diff
@@ -118,17 +137,104 @@ defmodule Phoenix.Socket.Transport do
...
+    * "heartbeat" events in the "phoenix" topic - should just emit
+      an OK reply
+    * "phx_join" on any topic - should join the topic
+    * "phx_leave" on any topic - should leave the topic
+
+  """
+  @callback handle_in({message :: term, opts :: keyword}, state) ::
+              {:ok, state}
+              | {:reply, status :: atom, {opcode :: atom, message :: term}, state}
+              | {:stop, reason :: term, state}
```
Comment (sasa1977) [inline lib/phoenix/socket/transport.ex]: Would it also be possible to support hibernation here and in `handle_info`?

File: lib/phoenix/channel/server.ex [hunk 5]
```diff
@@ -10,55 +10,91 @@ defmodule Phoenix.Channel.Server do
...
+          private: channel.__socket__(:private)
+      }
+
+    instrument = %{params: payload, socket: socket}
+
+    Phoenix.Endpoint.instrument socket, :phoenix_channel_join, instrument, fn ->
       ref = make_ref()
 
-      case GenServer.start_link(__MODULE__, {socket, auth_payload, self(), ref}) do
+      case Supervisor.start_child(socket.handler, [socket, payload, self(), ref]) do
```
Comment (sasa1977) [inline lib/phoenix/channel/server.ex]: Do I get this right that this PR also changes the process structure to introduce a dedicated supervisor for each socket connection?

File: lib/phoenix/channel/server.ex [hunk 6]
```diff
@@ -10,55 +10,91 @@ defmodule Phoenix.Channel.Server do
...
-      timeout ->
-        Process.exit(pid, :kill)
-        receive do
-          {:DOWN, ^ref, _, _, _} -> :ok
-        end
+    # otherwise the link in the transport will trigger it to crash.
+    pids_and_refs =
+      for pid <- pids do
+        ref = Process.monitor(pid)
+        GenServer.cast(pid, :close)
```
Comment (pdawczak) [inline lib/phoenix/channel/server.ex]: Interesting - why `cast` instead of `stop`?

File: lib/phoenix/socket.ex [hunk 7]
```diff
@@ -122,50 +123,285 @@ defmodule Phoenix.Socket do
...
+    :erlang.garbage_collect(self())
+    {:ok, state}
+  end
+
+  def __info__(_, state) do
+    {:ok, state}
+  end
+
+  def __terminate__(_reason, {%{channels_inverse: channels_inverse}, _socket}) do
+    Phoenix.Channel.Server.close(Map.keys(channels_inverse))
```
Comment (michalmuskala) [inline lib/phoenix/socket.ex]: Wouldn't it make sense to exit the supervisor for the channels here instead of manually closing the connections?

File: lib/phoenix/socket/transport.ex (hunk 4 above)
Comment (josevalim) [reply lib/phoenix/socket/transport.ex]: It would be possible - I don't plan to tackle it in this PR though.

File: lib/phoenix/channel/server.ex (hunk 5 above)
Comment (josevalim) [reply lib/phoenix/channel/server.ex]: Yes!

File: lib/phoenix/channel/server.ex (hunk 6 above)
Comment (josevalim) [reply lib/phoenix/channel/server.ex]: `stop` would crash if the process is already dead or dies meanwhile.

File: lib/phoenix/socket.ex (hunk 7 above)
Comment (josevalim) [reply lib/phoenix/socket.ex]: The channel supervisor is shared across all sockets.

Comment (josevalim) [reply lib/phoenix/socket.ex]: We could change it so each socket has its own supervisor but that would imply even more sockets per connection. The opposite concern though is that the supervisor we added may become a bottleneck. Unlikely though.

File: mix.exs (hunk 3 above)
Comment (josevalim) [reply mix.exs]: I don't think it was. I will check.

File: lib/phoenix/transports/long_poll.ex (hunk 2 above)
Comment (josevalim) [reply lib/phoenix/transports/long_poll.ex]: If we do that, we need a special formatter only for long polling. I want to share them, which means they have to return already encoded data.

File: lib/phoenix/channel/server.ex (hunk 1 above)
Comment (josevalim) [reply lib/phoenix/channel/server.ex]: Yes, we usually do it to say "i know this returns something meaningful, but i am sure i don't need it".

File: lib/phoenix/channel/server.ex (hunk 5 above)
Comment (sasa1977) [reply lib/phoenix/channel/server.ex]: Cool! So I guess that this might also fix #2466 then.

File: lib/phoenix/channel/server.ex (hunk 6 above)
Comment (sasa1977) [reply lib/phoenix/channel/server.ex]: Based on the usage below, the timeout semantics seems weird. If I read this right, it could happen that if you had say three channels which refuse to stop, and a timeout is say 5 seconds, this whole termination code might run for 15 seconds.

Comment (josevalim) [reply lib/phoenix/channel/server.ex]: Good catch! The current code is better than the previous because we at least close them all at the same time but this should be fixed and use a strategy similar to supervisors.

Comment (pdawczak) [reply lib/phoenix/channel/server.ex]: Thank you @josevalim - I certainly need to dig that one in a little bit more detail then!

File: lib/phoenix/transports/long_poll.ex [hunk 8]
```diff
@@ -261,18 +190,24 @@ defmodule Phoenix.Transports.LongPoll do
...
+    token = Phoenix.json_library().encode_to_iodata!(token)
+    data = [{"status", Integer.to_string(status)}, {"token", token}, {"messages", messages}]
+    send_json(conn, data)
   end
 
-  defp code_reload(conn, opts, endpoint) do
-    reload? = Keyword.get(opts, :code_reloader, endpoint.config(:code_reloader))
-    if reload?, do: Phoenix.CodeReloader.reload!(endpoint)
+  defp send_json(conn, json) do
+    data = for {k, v} <- json, do: [?", k, ?", ?:, v]
```
Comment (michalmuskala) [inline lib/phoenix/transports/long_poll.ex]: Are you sure the data does not contain anything that has to be escaped in JSON? Why not use the encoder?

Comment (josevalim) [reply lib/phoenix/transports/long_poll.ex]: Yes, I am because all of the calls happen on this module.

We cannot use the encoder because the data that comes from the socket is already encoded. Otherwise we would need to encode it twice.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
